### PR TITLE
WIP for GAWB-2450: validate MC expressions when validating/creating Submissions

### PR DESCRIPTION
**Pull request by jmthibault79**
- [x] **Submitter**: Include the JIRA issue number in the PR description
- [ ] **Submitter**: Check that the **Product Owner** has signed off on any user-facing changes
- [ ] **Submitter**: Make sure Swagger is updated if API changes
  - [ ] **...and Orchestration's Swagger too!**
- [ ] **Submitter**: If updating admin endpoints, also update [firecloud-admin-cli](https://github.com/broadinstitute/firecloud-admin-cli)
- [ ] **Submitter**: Check documentation and code comments. Add explanatory PR comments if helpful.
- [ ] **Submitter**: JIRA ticket checks:
  * Acceptance criteria exists and is met
  * Note any changes to implementation from the description
  * To Demo flag is set
  * Release Summary is filled out, if applicable
  * Add notes on how to QA
- [ ] **Submitter**: Update RC_XXX release ticket with any config or environment changes necessary
- [ ] **Submitter**: Database checks:
  * If PR includes new or changed db queries, include the explain plans in the description
  * Make sure liquibase is updated if appropriate
  * If doing a migration, take a backup of the
  [dev](https://console.cloud.google.com/sql/instances/terraform-qfarbdq3lrexxck5htofjs5z6m/backups?project=broad-dsde-dev&organizationId=548622027621)
  and
  [alpha](https://console.cloud.google.com/sql/instances/terraform-r4caezzc35c4tb7pgdhwkmme4y/backups?project=broad-dsde-alpha&organizationId=548622027621)
  DBs in Google Cloud Console
- [ ] **Submitter**: Update FISMA documentation if changes to:
  * Authentication
  * Authorization
  * Encryption
  * Audit trails
- [ ] Tell your tech lead (TL) that the PR exists if they want to look at it
- [ ] Anoint a lead reviewer (LR). **Assign PR to LR**
* Review cycle:
  * LR reviews
  * Rest of team may comment on PR at will
  * **LR assigns to submitter** for feedback fixes
  * Submitter rebases to develop again if necessary
  * Submitter makes further commits. DO NOT SQUASH
  * Submitter updates documentation as needed
  * Submitter **reassigns to LR** for further feedback
- [ ] **TL** sign off
- [ ] **LR** sign off
- [ ] **Assign to submitter** to finalize
- [ ] **Submitter**: Verify all tests go green, including CI tests
- [ ] **Submitter**: Squash commits and merge to develop
- [ ] **Submitter**: Delete branch after merge
- [ ] **Submitter**: **Test this change works on dev environment after deployment**. YOU own getting it fixed if dev isn't working for ANY reason!
- [ ] **Submitter**: Verify swagger UI on dev environment still works after deployment
- [ ] **Submitter**: Inform other teams of any API changes via Slack and/or email
- [ ] **Submitter**: Mark JIRA issue as resolved once this checklist is completed

### CODE REVIEW

**Comment by jmthibault79** [comment]
and a few other small changes

**Comment by jmthibault79** [comment]
TODO
* validate output expressions
* test in UI using Local stack
* write up future-work/GAWB-2444 doc
* refactor Input/Output Expression parsing to be more DRY
* use a `Validated` in `gatherInputs()` ?

File: core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/MethodConfigurationComponent.scala [hunk 1]
```diff
@@ -236,18 +234,6 @@ trait MethodConfigurationComponent {
       findByWorkspace(workspaceId).filterNot(_.deleted)
     }
 
-    private def findInputsByConfigId(configId: Long): MethodConfigurationInputQueryType = {
```
Comment (jmthibault79) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/MethodConfigurationComponent.scala:239]: unused

File: core/src/main/scala/org/broadinstitute/dsde/rawls/expressions/OutputExpression.scala [hunk 2]
```diff
@@ -1,39 +0,0 @@
-package org.broadinstitute.dsde.rawls.expressions
```
Comment (jmthibault79) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/expressions/OutputExpression.scala:1]: Moved to `model`

File: EXPRESSIONS.md [hunk 3]
```diff
@@ -46,7 +46,7 @@ Entity Attributes: ```control_bai```, ```case_bai```, ```output_vcf```, ```ref_p
...
+Input and output expressions will tell the method configuration what to associate the WDL variables with. When writing an expression, ```this``` refers to the root entity that is selected when configuring a method. To write an expression that binds to data associated with the root entity, use the ```this.<attribute_name>``` syntax (i.e. to access a reference FASTA associated with the example root entity (shown above), use ```this.ref_fasta```, where ```ref_fasta``` is the attribute name associated with your FASTA file pointer.) Input expressions may also dereference intermediate entities from the root entity. For example, to dereference an attribute called ```ref_fasta``` on the entity ```intermediate_entity```, use the expression  ```this.intermediate_entity.ref_fasta```. Output expressions do not have the ability to dereference intermediate entities. To bind data in the workspace, use the ```workspace.<attribute_name>``` syntax, which functions similarly to the ```this.``` syntax.  Output expressions can be blank, to indicate that the results of the workflow will not be bound.
```
Comment (jmthibault79) [inline EXPRESSIONS.md:49]: Added the last line

File: core/src/main/scala/org/broadinstitute/dsde/rawls/jobexec/SubmissionMonitorActor.scala [hunk 4]
```diff
@@ -378,7 +378,7 @@ trait SubmissionMonitor extends FutureSupport with LazyLogging with RawlsInstrum
         outputs.get(outputName) match {
           case None => Failure(new RawlsException(s"output named ${outputName} does not exist"))
           case Some(Right(uot: UnsupportedOutputType)) => Failure(new RawlsException(s"output named ${outputName} is not a supported type, received json u${uot.json.compactPrint}"))
-          case Some(Left(output)) => OutputExpression.build(outputExprStr, output)
```
Comment (jmthibault79) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/jobexec/SubmissionMonitorActor.scala:381]: `build()` no longer requires an output to bind to.  Use `bind()` to do that.

File: core/src/main/scala/org/broadinstitute/dsde/rawls/jobexec/SubmissionMonitorActor.scala [hunk 5]
```diff
@@ -396,7 +396,7 @@ trait SubmissionMonitor extends FutureSupport with LazyLogging with RawlsInstrum
   }
 
   def updateEntityAndWorkspace(entity: Entity, workspace: Workspace, workflowOutputs: Iterable[BoundOutputExpression]): (WorkflowEntityUpdate, Option[Workspace]) = {
-    val entityUpsert = workflowOutputs.collect({ case BoundOutputExpression(ThisEntityTarget, attrName, attr) => (attrName, attr) })
+    val entityUpsert = workflowOutputs.collect({ case BoundOutputExpression(EntityTarget, attrName, attr) => (attrName, attr) })
```
Comment (jmthibault79) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/jobexec/SubmissionMonitorActor.scala:399]: `EntityTarget` can be used by `InputExpression`s to refer to an entity other than `this`

File: core/src/test/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/MethodConfigurationComponentSpec.scala [hunk 6]
```diff
@@ -10,17 +10,16 @@ import org.broadinstitute.dsde.rawls.model._
...
 
     val methodConfig2 = MethodConfiguration(
       "ns",
       "config2",
       "sample",
-
-      Map("input.expression" -> AttributeString("this..wont.parse")),
-      Map("output.expression" -> AttributeString("output.expr")),
       Map("prereq.expression" -> AttributeString("prereq.expr")),
+      Map("input.expression" -> AttributeString("this.input_expr")),
```
Comment (jmthibault79) [inline core/src/test/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/MethodConfigurationComponentSpec.scala:21]: This change is not strictly necessary - it originally came from my local 2444 wip branch

File: core/src/test/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/TestDriverComponent.scala [hunk 7]
```diff
@@ -387,21 +387,20 @@ trait TestDriverComponent extends DriverComponent with DataAccess with DefaultIn
...
     val methodConfigWorkspaceUpdate = MethodConfiguration("ns", "testConfig1", "Sample", Map(), Map(), Map("o1" -> AttributeString("workspace.foo")), MethodRepoMethod("ns-config", "meth1", 1))
     val methodConfigWorkspaceLibraryUpdate = MethodConfiguration("ns", "testConfigLib", "Sample", Map(), Map(), Map("o1" -> AttributeString("workspace.library:foo")), MethodRepoMethod("ns-config", "meth1", 1))
     val methodConfigMissingOutputs = MethodConfiguration("ns", "testConfigMissingOutputs", "Sample", Map(), Map(), Map("some.workflow.output" -> AttributeString("this.might_not_be_here")), MethodRepoMethod("ns-config", "meth1", 1))
 
     val methodConfigValid = MethodConfiguration("dsde", "GoodMethodConfig", "Sample", prerequisites=Map.empty, inputs=Map("three_step.cgrep.pattern" -> AttributeString("this.name")), outputs=Map.empty, MethodRepoMethod("dsde", "three_step", 1))
-    val methodConfigUnparseable = MethodConfiguration("dsde", "UnparseableMethodConfig", "Sample", prerequisites=Map.empty, inputs=Map("three_step.cgrep.pattern" -> AttributeString("this..wont.parse")), outputs=Map.empty, MethodRepoMethod("dsde", "three_step", 1))
```
Comment (jmthibault79) [inline core/src/test/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/TestDriverComponent.scala:404]: Moved to the tests where it is used, so other tests don't have to worry about an invalid MC.  (Like above, originally from the other branch)

File: core/src/test/scala/org/broadinstitute/dsde/rawls/jobexec/SubmissionSpec.scala [hunk 8]
```diff
@@ -381,17 +381,19 @@ class SubmissionSpec(_system: ActorSystem) extends TestKit(_system) with FlatSpe
     }
   }
 
-  it should "return a successful Submission but no started workflows when given a method configuration with unparseable inputs" in withWorkspaceService { workspaceService =>
-    val submissionRq = SubmissionRequest("dsde", "UnparseableMethodConfig", "Individual", "indiv1", Some("this.sset.samples"), false)
-    val rqComplete = Await.result(workspaceService.createSubmission( testData.wsName, submissionRq ), Duration.Inf).asInstanceOf[RequestComplete[(StatusCode, SubmissionReport)]]
-    val (status, newSubmissionReport) = rqComplete.response
-    assertResult(StatusCodes.Created) {
-      status
+  it should "400 when given a method configuration with unparseable inputs" in withWorkspaceService { workspaceService =>
```
Comment (jmthibault79) [inline core/src/test/scala/org/broadinstitute/dsde/rawls/jobexec/SubmissionSpec.scala:384]: This is a behavior change so we may want to inform the API list

File: core/src/test/scala/org/broadinstitute/dsde/rawls/mock/RemoteServicesMockServer.scala [hunk 9]
```diff
@@ -41,19 +41,19 @@ class RemoteServicesMockServer(port:Int) extends RawlsTestUtils {
     val copyMethodConfigPath = "/configurations"
 
     val goodResult = AgoraEntity(Some("workspace_test"), Some("rawls_test_good"), Some(1), None, None, None, None,
-      Some("{\"name\":\"testConfig1\",\"workspaceName\":{\"namespace\":\"myNamespace\",\"name\":\"myWorkspace\"},\"methodRepoMethod\":{\"methodNamespace\":\"ns-config\",\"methodName\":\"meth1\",\"methodVersion\":1},\"methodRepoConfig\":{\"methodConfigNamespace\":\"ns\",\"methodConfigName\":\"meth1\",\"methodConfigVersion\":1},\"outputs\":{\"p1\":\"prereq expr\"},\"inputs\":{\"o1\":\"output expr\"},\"rootEntityType\":\"Sample\",\"prerequisites\":{\"i1\":\"input expr\"},\"namespace\":\"ns\"}"),
+      Some("""{"name":"testConfig1","workspaceName":{"namespace":"myNamespace","name":"myWorkspace"},"methodRepoMethod":{"methodNamespace":"ns-config","methodName":"meth1","methodVersion":1},"methodRepoConfig":{"methodConfigNamespace":"ns","methodConfigName":"meth1","methodConfigVersion":1},"outputs":{"o1":"this.output_expr"},"inputs":{"i1":"this.input_expr"},"rootEntityType":"Sample","prerequisites":{"i1":"input_expr"},"namespace":"ns"}"""),
```
Comment (jmthibault79) [inline core/src/test/scala/org/broadinstitute/dsde/rawls/mock/RemoteServicesMockServer.scala:44]: formatting only

File: core/src/test/scala/org/broadinstitute/dsde/rawls/webservice/MethodConfigApiServiceSpec.scala [hunk 10]
```diff
@@ -38,7 +38,7 @@ class MethodConfigApiServiceSpec extends ApiServiceSpec {
   }
 
   "MethodConfigApi" should "return 201 on create method configuration" in withTestDataApiServices { services =>
-    val newMethodConfig = MethodConfiguration("dsde", "testConfigNew", "samples", Map("ready" -> AttributeString("true")), Map("param1" -> AttributeString("foo")), Map("out" -> AttributeString("bar")),
+    val newMethodConfig = MethodConfiguration("dsde", "testConfigNew", "samples", Map("ready" -> AttributeString("true")), Map("param1" -> AttributeString("this.foo")), Map("out" -> AttributeString("this.bar")),
```
Comment (jmthibault79) [inline core/src/test/scala/org/broadinstitute/dsde/rawls/webservice/MethodConfigApiServiceSpec.scala:41]: Again - this comes from a different iteration where we enforce valid MC expressions on save

File: model/src/main/scala/org/broadinstitute/dsde/rawls/model/expressions/InputExpression.scala [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+  override def toString: String = {
+    val pathStr = attributePath map AttributeName.toDelimitedName mkString "."
+    target.root + pathStr
+  }
+}
+
+object TargetedInputExpression extends StringValidationUtils {
+  override implicit val errorReportSource = ErrorReportSource("rawls")
+
+  def tryParse(target: ExpressionTarget, expr: String): Try[TargetedInputExpression] = {
```
Comment (jmthibault79) [inline model/src/main/scala/org/broadinstitute/dsde/rawls/model/expressions/InputExpression.scala:23]: Very similar to the same in `TargetedOutputExpression`.  TODO: DRY

File: core/src/main/scala/org/broadinstitute/dsde/rawls/jobexec/MethodConfigResolver.scala [hunk 12]
```diff
@@ -57,19 +58,22 @@ object MethodConfigResolver {
...
-      val message =
-        if (missingInputs.nonEmpty)
-          if (extraInputs.nonEmpty)
-            "is missing definitions for these inputs: " + missingInputs.mkString(", ") + " and it has extraneous definitions for these inputs: " + extraInputs.mkString(", ")
-          else
-            "is missing definitions for these inputs: " + missingInputs.mkString(", ")
-        else
-          "has extraneous definitions for these inputs: " + extraInputs.mkString(", ")
-      throw new RawlsException(s"MethodConfiguration ${methodConfig.namespace}/${methodConfig.name} ${message}")
+    val extraInputs = methodConfig.inputs.keys.filterNot { agoraInputs.contains }
```
Comment (jmthibault79) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/jobexec/MethodConfigResolver.scala:65]: TODO: `Validated`?

File: model/src/test/scala/org/broadinstitute/dsde/rawls/model/expressions/ExpressionFixture.scala [hunk 13]
```diff
@@ -0,0 +1,77 @@
+package org.broadinstitute.dsde.rawls.model.expressions
+
+trait ExpressionFixture {
```
Comment (jmthibault79) [inline model/src/test/scala/org/broadinstitute/dsde/rawls/model/expressions/ExpressionFixture.scala:3]: Can you think of more test cases?

File: model/src/test/scala/org/broadinstitute/dsde/rawls/model/expressions/InputExpressionSpec.scala [hunk 14]
```diff
@@ -0,0 +1,58 @@
+package org.broadinstitute.dsde.rawls.model.expressions
+
+import org.broadinstitute.dsde.rawls.model.AttributeName
+import org.scalatest.{FlatSpec, Matchers}
+import spray.json.{JsArray, JsNumber, JsObject, JsString, JsTrue}
+
+import scala.util.{Failure, Success}
+
+class InputExpressionSpec extends FlatSpec with Matchers with ExpressionFixture {
```
Comment (jmthibault79) [inline model/src/test/scala/org/broadinstitute/dsde/rawls/model/expressions/InputExpressionSpec.scala:9]: TODO: add more of the test fixture cases explicitly?

**Comment by helgridly** [comment]
Have you seen `WorkspaceService.validateMCExpressions()`? Right now you're validating input expressions in gatherInputs, which is at submission-evaluation time, and validating output expressions at output-binding time. The latter is _certainly_ too late, as the job has finished running by then and the user is gonna be mad if it fails because they made a typo.

So: if you add your validation to `validateMCExpressions` that'll make sure that the user knows they've written junk in the MC because they'll see red on the MC page. That's the important bit. Note that this doesn't prevent the MC being saved to the database, it just returns a `ValidatedMethodConfiguration` with some errors in it.

For creating or validating submissions, what currently happens is that we skip calling `validateMCExpressions` in favour of simply trying to evaluate the expressions; if expression evaluation fails, the submissions insta-fails. This is where the current behaviour is out of whack, because the expression evaluator and `validateMCExpressions` don't entirely agree. Validating MCs and validating submissions should probably share code as much as possible; at the very least it's important that they behave the same way when it comes to expression syntax. (You'll potentially get _more_ errors at submission evaluation time as you're evaluating the expressions rather than parsing them.)

That said, I _don't_ think it's necessary for the expression parser(s) to explicitly support blank output expressions (or blank optional inputs) -- we can verify beforehand that it's okay for that input/output to be empty and not bother trying to evaluate it.

Does that make sense? Or have I missed the point entirely?

File: core/src/test/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/MethodConfigurationComponentSpec.scala [hunk 15]
```diff
@@ -10,17 +10,16 @@ import org.broadinstitute.dsde.rawls.model._
 class MethodConfigurationComponentSpec extends TestDriverComponentWithFlatSpecAndMatchers with RawlsTestUtils {
   import driver.api._
 
-  "MethodConfigurationComponenent" should "save and get a method config" in withDefaultTestDatabase {
```
Comment (helgridly) [inline core/src/test/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/MethodConfigurationComponentSpec.scala:13]: what a great typo

File: core/src/test/scala/org/broadinstitute/dsde/rawls/jobexec/SubmissionSpec.scala (hunk 8 above)
Comment (helgridly) [reply core/src/test/scala/org/broadinstitute/dsde/rawls/jobexec/SubmissionSpec.scala:384]: what's the motivation for the behaviour change here? (i think it's probably the right change, fwiw)

File: model/src/main/scala/org/broadinstitute/dsde/rawls/model/expressions/OutputExpression.scala [hunk 16]
```diff
@@ -0,0 +1,56 @@
...
+case object BlankOutputExpression extends OutputExpression {
+  override def toString: String = ""
+}
+case class TargetedOutputExpression(target: ExpressionTarget, attributeName: AttributeName) extends OutputExpression {
+  override def toString: String = s"${target.root}${AttributeName.toDelimitedName(attributeName)}"
+}
+
+sealed trait OutputExpressionBinding
+case object UnboundOutputExpression extends OutputExpressionBinding
+case class BoundOutputExpression(target: ExpressionTarget, attributeName: AttributeName, attribute: Attribute) extends OutputExpressionBinding
```
Comment (helgridly) [inline model/src/main/scala/org/broadinstitute/dsde/rawls/model/expressions/OutputExpression.scala:18]: can you talk some about these classes? the difference between building and binding? what they represent and their flow? i'm not super following here

File: model/src/main/scala/org/broadinstitute/dsde/rawls/model/expressions/InputExpression.scala [hunk 17]
```diff
@@ -0,0 +1,48 @@
...
+      if (rawNames.isEmpty || rawNames.exists(_.isEmpty)) throw new RawlsException(s"Invalid input expression: $expr")
+
+      val attributeNames = rawNames map { name =>
+        val attributeName = AttributeName.fromDelimitedName(name)
+        validateUserDefinedString(attributeName.name)
+        attributeName
+      }
+      TargetedInputExpression(target, attributeNames)
+    }
+    else Failure(new RawlsException(s"Invalid input expression: $expr"))
```
Comment (ansingh7115) [inline model/src/main/scala/org/broadinstitute/dsde/rawls/model/expressions/InputExpression.scala:35]: Maybe some comments about what constitutes valid vs invalid input and output expressions

File: model/src/main/scala/org/broadinstitute/dsde/rawls/model/expressions/OutputExpression.scala (hunk 16 above)
Comment (ansingh7115) [reply model/src/main/scala/org/broadinstitute/dsde/rawls/model/expressions/OutputExpression.scala:18]: +1

File: core/src/main/scala/org/broadinstitute/dsde/rawls/jobexec/MethodConfigResolver.scala [hunk 18]
```diff
@@ -57,19 +58,22 @@ object MethodConfigResolver {
...
+    val invalidInputs = methodConfig.inputs.collect {
+      case (name, AttributeString(expr)) if InputExpression.build(expr).isFailure => s"$name -> $expr"
```
Comment (ansingh7115) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/jobexec/MethodConfigResolver.scala:67]: Does the function `resolveInputsForEntities` right beneath this one require fiddling with?

**Comment by jmthibault79** [comment]
It does not make sense to review this PR in its current state.  Please read [the tech doc](https://docs.google.com/document/d/1t4rOeWo7wRTZJlZGfwaZsLEyYvcxFLtSx7Q5eTYsMAg/edit#heading=h.22whn74qtfyt) and I'll get back to this soon.  Thanks :)

**Pull request closed without merging**